prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -0,0 +1,136 @@
+            return pd.NA
+    distance_to_subs = matching_subs.apply(
+        lambda x: haversine((x.LATITUDE, x.LONGITUDE), (generator.lat, generator.lon)),
+        axis=1,
+    )
+    return distance_to_subs.idxmin()
+
+
+def map_generator_to_bus_by_sub(generator, bus_groupby):
+    """Determine a likely bus for a generator to be connected to, based on the bus with
+    the lowest voltage within the generator's specified substation.
+
+    :param pandas.Series generator: one generating unit from data frame.
+    :param pandas.GroupBy bus_groupby: data frame of buses, grouped by substation ID.
+    :return: (*int/pd.NA*) -- bus ID if the generator has a substation ID, else pd.NA.
+    """
+    if pd.isna(generator.sub_id):
+        return pd.NA
+    else:
+        return bus_groupby.get_group(generator.sub_id)["baseKV"].idxmin()
+
+
+def build_plant(bus, substations):
+    """Use source data on generating units from EIA/EPA, along with transmission network
+    data, to produce a plant data frame.
+
+    :param pandas.DataFrame bus: data frame of buses, to be used within
+        :func:`map_generator_to_bus_by_sub`.
+    :param pandas.DataFrame substations: data frame of substations.
+    :return: (*pandas.DataFrame*) -- data frame of generator data.
+    """
+    # Initial loading
+    generators = load.get_eia_form_860(const.blob_paths["eia_form860_2019_generator"])
+    plants = load.get_eia_form_860(const.blob_paths["eia_form860_2019_plant"])
+
+    # Data interpretation
+    plants = plants.set_index("Plant Code")
+    plants["Latitude"] = plants["Latitude"].map(floatify)
+    plants["Longitude"] = plants["Longitude"].map(floatify)
+    for col in ["Summer Capacity (MW)", "Winter Capacity (MW)", "Minimum Load (MW)"]:
+        generators[col] = generators[col].map(floatify)
+
+    # Filtering / Grouping
+    generators = generators.query(
+        "Technology not in @const.eia_storage_gen_types"
+    ).copy()
+    bus_groupby = bus.groupby(bus["sub_id"].astype(int))
+    # Filter substations with no buses
+    substations = substations.loc[set(bus_groupby.groups.keys())]
+    substation_groupby = substations.groupby(["interconnect", "ZIP"])
+
+    # Add information
+    generators["interconnect"] = (
+        generators["Plant Code"]
+        .map(plants["NERC Region"])
+        .map(const.nercregion2interconnect)
+    )
+    generators["lat"] = generators["Plant Code"].map(plants["Latitude"])
+    generators["lon"] = generators["Plant Code"].map(plants["Longitude"])
+    generators["ZIP"] = generators["Plant Code"].map(plants["Zip"])
+    generators["sub_id"] = generators.apply(
+        lambda x: map_generator_to_sub_by_location(x, substation_groupby), axis=1
+    )
+    generators["bus_id"] = generators.apply(
+        lambda x: map_generator_to_bus_by_sub(x, bus_groupby), axis=1
+    )
+    generators["Pmax"] = generators[
+        ["Summer Capacity (MW)", "Winter Capacity (MW)"]
+    ].max(axis=1)
```

Comment:
Is that preferable than to take the average?

Reply:
Good question. I erred on the side of creating a more-feasible system. Ideally we would have the ability to have seasonal capacities within PowerSimData/REISE.jl.

Reply:
Sounds good. I guess there is no seasonal difference for nuclear, natural gas and coal but could be significant for hydro, solar and wind

Reply:
Hydro, solar, and wind will be controlled by the profiles, but thermal generators can have also different net capacities in summer and winter because of ambient temperature differences, cooling water temperatures, etc.

Reply:
Ah. Thanks for the explanation